data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/ProcessWorker.java

```diff
@@ -44,41 +45,73 @@ public ProcessWorker(
     @Override
     public void run() {
         try {
-            do {
-                final Map.Entry<Collection, CheckpointState> readResult = readBuffer.read(pipeline.getReadBatchTimeoutInMillis());
-                Collection records = readResult.getKey();
-                final CheckpointState checkpointState = readResult.getValue();
-                //TODO Hacky way to avoid logging continuously - Will be removed as part of metrics implementation
-                if (records.isEmpty()) {
-                    if(!isEmptyRecordsLogged) {
-                        LOG.info(" {} Worker: No records received from buffer", pipeline.getName());
-                        isEmptyRecordsLogged = true;
-                    }
-                } else {
-                    LOG.info(" {} Worker: Processing {} records from buffer", pipeline.getName(), records.size());
-                }
-                //Should Empty list from buffer should be sent to the processors? For now sending as the Stateful processors expects it.
-                for (final Processor processor : processors) {
-                    records = processor.execute(records);
-                }
-                if (!records.isEmpty()) {
-                    postToSink(records);
-                }
-                // Checkpoint the current batch read from the buffer after being processed by processors and sinks.
-                readBuffer.checkpoint(checkpointState);
-            } while (!shouldStop());
+            // Phase 1 - execute until stop requested and buffers drained
+            while (!pipeline.isStopRequested()) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 1 complete.");
+
+            // Phase 2 - execute until read buffer and peer forwarder buffer are empty
+            LOG.info("Beginning processor shutdown phase 2, iterating until buffers empty.");
+            while (!areBuffersEmpty()) {
+                doRun();
```

Comment:
There's a chance this phase never ends if buffers are never getting empty.

Reply:
Yes, the amount of time given for this to complete is controlled by the [ExecutorService shutdown here](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java#L195). Users can configure the `processorShutdownTimeout` field in their `data-prepper-config.yaml` to control the max duration to wait for the shutdown process to complete before force terminating